text/0000-introduce-a-new-signing-mechanism.md

```diff
@@ -0,0 +1,568 @@
+
+1. Create a key pair: a private key for signing and a public key for adding to a public certificate.
+
+1. Sign the email address' SHA256 digest with the private key to prove possession of the private key. We will refer to this as the `proof` in the next step.
+
+1. Request a code signing certificate from Fulcio. Use the id token as the bearer token. The following example uses some pseudocode for brevity:
+
+   ```
+   POST https://fulcio.sigstore.dev/api/v1/signingCert
+   Authorization: "Bearer <the JWT>"
+   {
+       publicKey: {
+         content: Base64.encode64(pub_key.to_der),
+         algorithm: "ecdsa",
+       },
+       signedEmailAddress: Base64.encode64(proof),
+   }
+   ```
+
+   The response is a PEM-encoded pair of X.509 certificates: the code signing certificate with the user's email address as the subject alternative name, and Fulcio's own issuing certificate:
+   
+   ```
+   -----BEGIN CERTIFICATE-----
+   MIIDdjCCAvygAwIBAgITRT47a4MaF80DGP/Rcbt5eis4RjAKBggqhkjOPQQDAzAq
+   MRUwEwYDVQQKEwxzaWdzdG9yZS5kZXYxETAPBgNVBAMTCHNpZ3N0b3JlMB4XDTIy
+   [rest of signing certificate]
+   -----END CERTIFICATE-----
+   
+   -----BEGIN CERTIFICATE-----
+   MIIB+DCCAX6gAwIBAgITNVkDZoCiofPDsy7dfm6geLbuhzAKBggqhkjOPQQDAzAq
+   MRUwEwYDVQQKEwxzaWdzdG9yZS5kZXYxETAPBgNVBAMTCHNpZ3N0b3JlMB4XDTIx
+   [rest of issuing (root) certificate]
+   -----END CERTIFICATE-----
+   ```
+
+1. Compute a SHA256 digest of the entire .gem file's contents, then sign the digest using the private signing key.
+
+1. Submit the signature and public signing certificate to Rekor before the signing certificate expires (example uses pseudocode):
+
+   ```
+   POST https://rekor.sigstore.dev/api/v1/log/entries
+   {
+    kind: "rekord",
+    apiVersion: "0.0.1",
+    spec: {
+      signature: {
+        format: "x509",
+        content: Base64.encode64(signature),
+        publicKey: {
+          content: Base64.encode64(cert_chain_pem),
+        },
+      },
+      data: {
+        content: Base64.encode64(gem_contents),
+        hash: {
+          algorithm: "sha256",
+          value: gem_digest,
+        },
+      },
+    },
+   }
+   ```
+
+The gem signature and certificate chain are stored together in an immutable Rekor log entry, such as https://rekor.sigstore.dev/api/v1/log/entries/e66b75a340d2cf240f92cc0fd7f77c3ea0855c7eb5e3e9e2ae873567d9491fdc. In the signature verification flow, we will be able to locate this entry by searching Rekor using the .gem file's digest.
+
+### The verifying flow
+
+The signature verification flow occurs at gem installation time. In our proposed Phase 1, verification is opt-in, and the signature entry is only hosted on Rekor itself. By Phase 2, we anticipate that rubygems.org will keep a copy of the signature as well, and that clients will download it alongside the gem itself, decoupling the install experience from any Rekor queries on the happy path.
```

Comment:
I'd be interested in more details here about what exactly would be kept alongside the gem, and in what format. Is this just the `.Body.RekordObj.signature` field from the transparency log, as JSON? Something else?

Reply:
Good question! I envision we'd keep the entire log entry response body from Rekor, as a JSON file. So it's the response from `POST /log/entries/retrieve`, in verification step 2.

Most of the interesting signature bits live in the body, i.e. the _rekord_.  The log entry itself contains the `integratedTime` field, which is the point in time where the certificate chain must be valid. A client would need to know this before it accepts the cert chain.

It also contains verification material for the entry. We're most interested in the signedEntryTimestamp,  a signed digest of the entry's body, integratedTime, and logIndex(?).  Anyone with Rekor's verification key can verify the authenticity of the log entry without needing to phone home.

TODO: insert dad joke about _s.E.T. no phone home_

I have no opinions on the name of this sig file.  A gem could be signed multiple times, so it's not as straightforward as `gemname.version.sig`.

Reply:
Agreed, I am similarly interested in avoiding 'phoning home' to rekor. And I agree that whatever happens here should be able to support multiple signatures, but also be able to be tied to the original artifact.

Maybe for an artifact like `gemname-1.0.0.gem`, something like `<artifact filename>.<some unique value>.sig`, where `<some unique value>` is one of:

* the Rekor `LogIndex`?
  * e.g. `gemname-1.0.0.gem.1218056.sig`
* the Rekor `UUID`/`LogID`? (probably too long)
  * e.g. `gemname-1.0.0.gem.06a53257a4cb3e4d0589b117de8b14712927ad5a73da5279f00d4d88a5adfb2d.sig`
  * e.g. `gemname-1.0.0.gem.c0d23d6ad406973f9559f3ba2d1ca01f84147d8ffc5b8445c224f98b9591801d.sig`
* the Rekor `sha256` hash? (also probably too long, but...)
  * e.g. `gemname-1.0.0.gem.98ea6e4f216f2fb4b69fff9b3a44842c38686ca685f3f55dc48c5d3fb1107be4.sig`
* the first N characters of the Rekor `sha256` hash, similar to git short hashes?
  * e.g. `gemname-1.0.0.gem.98ea6e4.sig`
  
I think I like the last one the best.

Reply:
Maybe multiple signatures is one of those things that should be punted to future versions of the RFC which may include TUF, which is designed for problems such as these?

Reply:
> the first N characters of the Rekor sha256 hash, similar to git short hashes?
> e.g. gemname-1.0.0.gem.98ea6e4.sig

I also like this approach. Since the rekor hash is the hash of the signed bytes/gem, it's trivial to derive the expected name of the signature from just the gem. This won't work for multiple signatures, so possibly a monotonically increasing version, `gemname-1.0.0.gem.<hash>.<version>.sig`?

I also wanted to call out that for verifying a persisted bundle instead of hitting the log, we'll need to make sure we verify that the offline Rekor bundle is actually linked to the gem. For example, [Cosign](https://github.com/sigstore/cosign/blob/2f5e1ffa0fe0c1104355e18a1bc77d91ac301a85/pkg/cosign/verify.go#L623-L629) does this on verification by calculating the hash of the payload and comparing it to the hash in the entry. This is necessary to make sure an arbitrary Rekor entry isn't bundled with the Gem.